### PR TITLE
ncm-opennebula: Use resource dictionaries instead of lists

### PR DESCRIPTION
**Pull request by alvarosimon**
This PR changes OpenNebula users/groups/vnets/datastores resources schema to use dictionaries

* Fix: #834 : Change some schema resource types to dictionaries instead lists

### CODE REVIEW

File: ncm-opennebula/src/main/resources/user.tt [hunk 1]
```diff
@@ -1,13 +1,14 @@
...
-[%         CASE 'user' -%]
-[%             IF pair.value != 'oneadmin' -%]
+[% FOR pool IN user.pairs -%]
+[%     FOR pair IN pool.value.pairs %]
+[%-        SWITCH pair.key -%]
+[%             CASE 'ssh_public_key' -%]
+[%                 pair.key FILTER upper %] = "
+[%-                 FOREACH item IN pair.value -%]
+[%                      item %]
+[%                  END -%]"
```
Comment (stdweird) [inline ncm-opennebula/src/main/resources/user.tt]: the default case is missing?

File: ncm-opennebula/src/main/resources/group.tt [hunk 2]
```diff
@@ -1,10 +1,11 @@
-[% FOR pair IN group.pairs -%]
-[%-    SWITCH pair.key -%]
-[%         CASE 'description' -%]
-[%             pair.key FILTER upper %] = "[% pair.value %]"
-[%         CASE 'group' -%]
-[%             IF pair.value != 'oneadmin' -%]
+[% FOR pool IN group.pairs -%]
+[%     FOR pair IN pool.value.pairs %]
+[%-        SWITCH pair.key -%]
+[%             CASE 'description' -%]
```
Comment (stdweird) [inline ncm-opennebula/src/main/resources/group.tt]: default case is missing?

File: ncm-opennebula/src/test/resources/one.pan [hunk 3]
```diff
@@ -35,93 +35,81 @@ prefix "/software/components/opennebula/kvmrc";
...
-    ),
-    dict(
-        "name", "nfs",
-        "datastore_capacity_check", true,
-        "ds_mad", "fs",
-        "tm_mad", "shared",
-        "type", "IMAGE_DS",
-    ),
+"datastores" = dict(
+        "ceph.altaria", dict(
```
Comment (stdweird) [inline ncm-opennebula/src/test/resources/one.pan]: 4 space indent

File: ncm-opennebula/src/main/resources/datastore.tt [hunk 4]
```diff
@@ -1,23 +1,26 @@
...
-[%                  CASE booleans -%]
- "[%                    pair.value ? "yes" : "no" %]"
-[%                  CASE integer -%]
- [%                     pair.value %]
-[%                  CASE -%]
- "[%                    pair.value %]"
-[%              END -%]
-[%      END -%]
+[%        IF pair.key == 'ceph_user_key' -%]
+[%             -%]
```
Comment (stdweird) [inline ncm-opennebula/src/main/resources/datastore.tt]: ?

File: ncm-opennebula/src/main/resources/user.tt (hunk 1 above)
Comment (alvarosimon) [inline ncm-opennebula/src/main/resources/user.tt]: That is done on purpose, we don't want to include some values in the template as example the user's password.

File: ncm-opennebula/src/main/resources/group.tt (hunk 2 above)
Comment (alvarosimon) [inline ncm-opennebula/src/main/resources/group.tt]: Same reason than the user case, maybe we can replace the `SWITCH` by `IF` statements.. but I'm not sure if we will need more conditions in the future

File: ncm-opennebula/src/main/resources/datastore.tt (hunk 4 above)
Comment (alvarosimon) [inline ncm-opennebula/src/main/resources/datastore.tt]: `ceph_user_key` is not included in the template for security reasons, only the Ceph UUID, but now reading the tt file it makes more sense to include this as a `SWITCH` case as we do for the users... I will change that

**Comment by piojo** [comment]
Why this change to dicts? What is "broken" with the old schema?

**Comment by alvarosimon** [comment]
@piojo nothing is broken in the current schema but with the dictionaries we can set the values just using the paths like `opennebula/user/username/ssh` or `opennebula/vnet/pool1/..` so it simplifies the configuration if you have many resources. This feature was requested by @stdweird, the only problem is the backwards incompatibility

**Comment by piojo** [comment]
Can you please paste here an example of an old template and a new one to illustrate the change the folks at ULB/VUB will need to do?

**Comment by alvarosimon** [comment]
@piojo  sure!
Here is how it looks the current template:

```

prefix "/software/components/opennebula";

"vnets" = list(
    dict(
        "name", "altaria.os",
        "bridge", "br100",
        "gateway", "10.141.3.250",
        "dns", "10.141.3.250",
        "network_mask", "255.255.0.0"
    ),
    dict(
        "name", "altaria.vsc",
        "bridge", "br101",
        "gateway", "10.141.3.250",
        "dns", "10.141.3.250",
        "network_mask", "255.255.0.0"
    ),
    dict(
        "name", "pool.altaria.os",
        "bridge", "br100",
        "bridge_ovs", "ovsbr0",
        "gateway", "10.141.3.250",
        "dns", "10.141.3.250",
        "network_mask", "255.255.0.0",
        "vlan", true,
        "vlan_id", 0,
        "vn_mad", "ovswitch",
        "ar", dict(
                    "type", "IP4",
                    "ip", "10.141.14.100",
                    "size", 29
        ),
    ),
);

"datastores" = list(
    dict(
        "name", "ceph.altaria",
        "bridge_list", list("hyp004.cubone.os"),
        "ceph_host", list("ceph001.cubone.os","ceph002.cubone.os","ceph003.cubone.os"),
        "ceph_secret", "8371ae8a-386d-44d7-a228-c42de4259c6e",
        "ceph_user", "libvirt",
        "datastore_capacity_check", true,
        "ceph_user_key", "xxxxxxxx",
        "pool_name", "one",
        "type", "IMAGE_DS"
    ),
    dict(
        "name", "nfs",
        "datastore_capacity_check", true,
        "ds_mad", "fs",
        "tm_mad", "shared",
        "type", "IMAGE_DS",
    ),
);

"groups" = list(
    dict(
        "group", "gvo01",
        "description", "gvo01 group managed by quattor",
    ),
);

"users" = list(
    dict(
        "user", "lsimngar",
        "password", "my_fancy_pass",
        "ssh_public_key", list(
            'ssh-dss A............'
            ),
        "quattor", 1,
        "group", "oneadmin",
    ),
    dict(
        "user", "stdweird",
        "password", "another_fancy_pass",
        "quattor", 1
    ),
    dict(
        "user", "serveradmin",
        "password", "yet_another_fancy_pass",
    ),
    dict(
        "user", "oneadmin",
        "ssh_public_key", list(
            'ssh-dss AAAAB3NzaC1yc2EAAAADAQABAAABAQDI4gvhOpwKbukZP/Tht/GmKcRCBHGn8JadVlgb9U6O/EP/hR1KLDbKY7KVjVOlUcvfawn44SIGsmKCzehYJV2s/XU1QSaaLrjB7n+vfOyj1C3EgzfZcMOHvL51xPuSgIoKd9oER/63B/pUV/BEZK5LEC06O1LgAjwLy2DrHNN3cQdnTbxQ4vM5ggDb/BC+DyRYlN5NG74VFguVQmoqMPA8FYXBvT/bBvIAZFw7piZIQFd6C803dtG61234 another@laptop'
        ),
    ),
);
```

And here how it looks like the same template but using dictionaries (only datastores,vnets,users and groups are affected by this change):

```
prefix "/software/components/opennebula";

"vnets" = dict(
    "altaria.os", dict(
            "bridge", "br100",
            "gateway", "10.141.3.250",
            "dns", "10.141.3.250",
            "network_mask", "255.255.0.0"
        ),
    "altaria.vsc", dict(
            "bridge", "br101",
            "gateway", "10.141.3.250",
            "dns", "10.141.3.250",
            "network_mask", "255.255.0.0"
        ),
    "pool.altaria.os", dict(
            "bridge", "br100",
            "bridge_ovs", "ovsbr0",
            "gateway", "10.141.3.250",
            "dns", "10.141.3.250",
            "network_mask", "255.255.0.0",
            "vlan", true,
            "vlan_id", 0,
            "vn_mad", "ovswitch",
            "ar", dict(
                    "type", "IP4",
                    "ip", "10.141.14.100",
                    "size", 29
            ),
        ),
);

"datastores" = dict(
        "ceph.altaria", dict(
            "bridge_list", list("hyp004.cubone.os"),
            "ceph_host", list("ceph001.cubone.os","ceph002.cubone.os","ceph003.cubone.os"),
            "ceph_secret", "8371ae8a-386d-44d7-a228-c42de4259c6e",
            "ceph_user", "libvirt",
            "datastore_capacity_check", true,
            "ceph_user_key", "AQCGZr1TeFUBMRBBHExosSnNXvlhuKexxcczpw==",
            "pool_name", "one",
            "type", "IMAGE_DS"
        ),
        "nfs", dict(
            "datastore_capacity_check", true,
            "ds_mad", "fs",
            "tm_mad", "shared",
            "type", "IMAGE_DS",
        ),
);

"groups" = dict(
        "gvo01", dict(
            "description", "gvo01 group managed by quattor",
        ),
);

"users" = dict(
        "lsimngar", dict(
            "password", "my_fancy_pass",
            "ssh_public_key", list(
                'ssh-dss A...'
            ),
            "group", "oneadmin",
        ),
        "stdweird", dict(
            "password", "another_fancy_pass",
        ),
        "serveradmin", dict(
            "password", "yet_another_fancy_pass",
        ),
        "oneadmin", dict(
            "ssh_public_key", list(
                'ssh-dss AAAAB3NzaC1yc2EAAAADAQABAAABAQDI4gvhOpwKbukZP/Tht/GmKcRCBHGn8JadVlgb9U6O/EP/hR1KLDbKY7KVjVOlUcvfawn44SIGsmKCzehYJV2s/XU1QSaaLrjB7n+vfOyj1C3EgzfZcMOHvL51xPuSgIoKd9oER/63B/pUV/BEZK5LEC06O1LgAjwLy2DrHNN3cQdnTbxQ4vM5ggDb/BC+DyRYlN5NG74VFguVQmoqMPA8FYXBvT/bBvIAZFw7piZIQFd6C803dtG61234 another@laptop'
            ),
        ),
);



```

**Pull request merged**